File: cli/yakuza_rogue_editor.py
import os
import json
import gzip
import sys

try:
    from Crypto.Cipher import AES
    from Crypto.Protocol.KDF import PBKDF2
    from Crypto.Hash import SHA1
    from Crypto.Random import get_random_bytes
except ImportError:
    print("Error: 'pycryptodome' package is required.")
    print("Please install it by running: pip install pycryptodome")
    sys.exit(1)

# The password for Yakuza Rogue: Yokohama massage parlor chapter / Dojo NTR save files
PASSWORD = "wanzg!1f**k"

def decrypt_es3(file_path):
    if not os.path.exists(file_path):
        print(f"Error: File '{file_path}' does not exist.")
        return None

    with open(file_path, "rb") as f:
        data = f.read()

    if len(data) < 16:
        print("Error: Invalid ES3 file (too short).")
        return None

    iv = data[:16]
    encrypted_data = data[16:]

    # Key derivation using PBKDF2 (SHA1, 100 iterations, 16 bytes key length)
    key = PBKDF2(PASSWORD.encode("utf-8"), iv, dkLen=16, count=100, hmac_hash_module=SHA1)
    
    cipher = AES.new(key, AES.MODE_CBC, iv)
    decrypted = cipher.decrypt(encrypted_data)

    # Remove PKCS7 padding
    padding_len = decrypted[-1]
    if 0 < padding_len <= 16:
        decrypted = decrypted[:-padding_len]

    # Check if data is gzipped
    is_gzipped = decrypted[0] == 0x1F and decrypted[1] == 0x8B
    if is_gzipped:
        try:
            decrypted = gzip.decompress(decrypted)
        except Exception as e:
            print("Warning: Failed to decompress gzipped data, returning raw decrypted data.", e)

    return decrypted.decode("utf-8", errors="ignore"), is_gzipped

def encrypt_es3(text_data, out_path, use_gzip=False):
    data = text_data.encode("utf-8")
    if use_gzip:
        data = gzip.compress(data)

    # PKCS7 Padding
    padding_len = 16 - (len(data) % 16)
    data += bytes([padding_len] * padding_len)

    iv = get_random_bytes(16)
    key = PBKDF2(PASSWORD.encode("utf-8"), iv, dkLen=16, count=100, hmac_hash_module=SHA1)

    cipher = AES.new(key, AES.MODE_CBC, iv)
    encrypted = cipher.encrypt(data)

    with open(out_path, "wb") as f:
        f.write(iv + encrypted)

def main():
    print("=" * 60)
    print(" YAKUZA ROGUE: YOKOHAMA MASSAGE PARLOR SAVE GAME EDITOR ")
    print("=" * 60)

    # Search for save file in default locations
    local_appdata = os.getenv("LOCALAPPDATA", "")
    default_dirs = [
        os.path.join(local_appdata + "Low", "Yakuza Rogue_ Yokohama massage parlor chapter", "SaveGames"),
        r"D:\SteamLibrary\steamapps\common\Massage Parlor\MassageShop_Data\game_data",
        r"C:\Program Files (x86)\Steam\steamapps\common\Massage Parlor\MassageShop_Data\game_data"
    ]
    
    save_path = None
    saves_found = []
    
    for d in default_dirs:
        if os.path.exists(d):
            saves = [os.path.join(d, f) for f in os.listdir(d) if f.endswith(".es3")]
            saves_found.extend(saves)
            
    if saves_found:
        print("\nFound save games:")
        for idx, s in enumerate(saves_found):
            print(f" [{idx + 1}] {s}")
        print(" [M] Manual path entry")
        
        choice = input("\nSelect a save file to edit: ").strip()
        if choice.lower() == 'm':
            save_path = input("Enter full path to save file (.es3): ").strip().strip('"')
        else:
            try:
                idx = int(choice) - 1
                if 0 <= idx < len(saves_found):
                    save_path = saves_found[idx]
                else:
                    print("Invalid choice.")
                    return
            except ValueError:
                print("Invalid choice.")
                return
    else:
        save_path = input("\nNo saves found automatically. Enter full path to save file (.es3): ").strip().strip('"')

    if not save_path or not os.path.exists(save_path):
        print("Invalid file path.")
        return

    # Backup original file
    backup_path = save_path + ".bak"
    if not os.path.exists(backup_path):
        import shutil
        shutil.copy2(save_path, backup_path)
        print(f"\n[+] Created backup of original save file at: {backup_path}")

    # Decrypt
    print("\nDecrypting save file...")
    decrypted_str, was_gzipped = decrypt_es3(save_path)
    if not decrypted_str:
        print("[-] Decryption failed. Please make sure the file is valid.")
        return

    try:
        save_json = json.loads(decrypted_str)
    except json.JSONDecodeError:
        print("[-] Decrypted content is not valid JSON. Save file might be corrupted or using a different password.")
        print(decrypted_str[:100])
        return

    print("[+] Decrypted successfully!")
    
    # Edit values
    modified = False
    
    # Check for known variables (Money, etc.)
    # In Yakuza Rogue / Dojo NTR, Money is often stored inside a specific JSON structure.
    # Let's inspect keys
    print("\n--- Save File Keys ---")
    keys_list = list(save_json.keys())
    for k in keys_list[:15]:
        val = save_json[k]
        print(f" - {k}: {val}")
    if len(keys_list) > 15:
        print(f" ... and {len(keys_list) - 15} more keys.")
        
    print("\n=== Editing Menu ===")
    print(" [1] Set Money to 9,999,999")
    print(" [2] Set task_fail_count to 0 (Reset mission failures)")
    print(" [3] Custom JSON Modification (advanced)")
    print(" [4] Just Export Decrypted JSON to text file")
    
    menu_choice = input("\nChoose an option: ").strip()
    
    if menu_choice == "1":
        # Search for money keys
        money_keys = [k for k in save_json if "money" in k.lower() or "gold" in k.lower() or "cash" in k.lower()]
        if money_keys:
            for k in money_keys:
                print(f"Changing {k} from {save_json[k]} to 9999999")
                save_json[k] = 9999999
            modified = True
        else:
            # Let's prompt user to input key
            key = input("Money key not found automatically. Enter key name (e.g. 'store_money' or 'money'): ").strip()
            if key in save_json:
                save_json[key] = 9999999
                modified = True
            else:
                print(f"Key '{key}' not found in save.")
                
    elif menu_choice == "2":
        fail_keys = [k for k in save_json if "fail" in k.lower()]
        if fail_keys:
            for k in fail_keys:
                print(f"Changing {k} from {save_json[k]} to 0")
                save_json[k] = 0
            modified = True
        else:
            key = input("Failure key not found automatically. Enter key name (e.g. 'task_fail_count'): ").strip()
            if key in save_json:
                save_json[key] = 0
                modified = True
            else:
                print(f"Key '{key}' not found in save.")
                
    elif menu_choice == "3":
        key = input("Enter the key name to modify: ").strip()
        if key in save_json:
            print(f"Current value: {save_json[key]} (type: {type(save_json[key])})")
            new_val = input("Enter new value: ").strip()
            # Try to convert to int/float/bool if possible
            if new_val.lower() == 'true':
                new_val = True
            elif new_val.lower() == 'false':
                new_val = False
            else:
                try:
                    if '.' in new_val:
                        new_val = float(new_val)
                    else:
                        new_val = int(new_val)
                except ValueError:
                    pass # Keep as string
            save_json[key] = new_val
            modified = True
        else:
            print(f"Key '{key}' not found.")
            
    elif menu_choice == "4":
        txt_path = save_path + "_decrypted.json"
        with open(txt_path, "w", encoding="utf-8") as f:
            json.dump(save_json, f, indent=4)
        print(f"[+] Decrypted JSON exported to: {txt_path}")
        print("You can edit this file manually and then run this script again to re-encrypt it.")
        return

    if modified:
        print("\nRe-encrypting and saving file...")
        new_text = json.dumps(save_json)
        encrypt_es3(new_text, save_path, use_gzip=was_gzipped)
        print("[+] Save file updated successfully! Enjoy!")

if __name__ == "__main__":
    main()

File: cli/yakuza_rogue_editor.js
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const zlib = require('zlib');
const readline = require('readline');

const PASSWORD = "wanzg!1f**k";

function preprocess_es3_json(jsonStr) {
    return jsonStr.replace(/([{,]\s*)([0-9]+)\s*:/g, '$1"$2":');
}

function postprocess_es3_json(jsonStr) {
    return jsonStr.replace(/([{,]\s*)"([0-9]+)"\s*:/g, '$1$2:');
}


function decrypt_es3(data, password) {
    try {
        const iv = data.subarray(0, 16);
        // Key derivation: pbkdf2 with sha1, 100 iterations, 16 bytes key size
        const key = crypto.pbkdf2Sync(password, iv, 100, 16, 'sha1');
        const decipher = crypto.createDecipheriv('aes-128-cbc', key, iv);
        
        let decrypted = Buffer.concat([decipher.update(data.subarray(16)), decipher.final()]);
        
        let wasGunzipped = false;
        // Check for gzip magic bytes 1F 8B
        if (decrypted[0] === 0x1F && decrypted[1] === 0x8B) {
            wasGunzipped = true;
            decrypted = zlib.gunzipSync(decrypted);
        }
        
        return { wasGunzipped, text: decrypted.toString('utf8'), success: true };
    } catch (e) {
        return { success: false, error: e.message };
    }
}

function encrypt_es3(textData, password, wasGunzipped) {
    let data = Buffer.from(textData, 'utf8');
    if (wasGunzipped) {
        data = zlib.gzipSync(data);
    }
    
    const iv = crypto.randomBytes(16);
    const key = crypto.pbkdf2Sync(password, iv, 100, 16, 'sha1');
    const cipher = crypto.createCipheriv('aes-128-cbc', key, iv);
    
    const encrypted = Buffer.concat([cipher.update(data), cipher.final()]);
    return Buffer.concat([iv, encrypted]);
}

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const question = (query) => new Promise((resolve) => rl.question(query, resolve));

async function main() {
    console.log("============================================================");
    console.log(" YAKUZA ROGUE: YOKOHAMA MASSAGE PARLOR SAVE GAME EDITOR ");
    console.log("============================================================");

    const localAppData = process.env.LOCALAPPDATA || "";
    const defaultDirs = [
        path.join(localAppData + "Low", "Yakuza Rogue_ Yokohama massage parlor chapter", "SaveGames"),
        "D:\\SteamLibrary\\steamapps\\common\\Massage Parlor\\MassageShop_Data\\game_data",
        "C:\\Program Files (x86)\\Steam\\steamapps\\common\\Massage Parlor\\MassageShop_Data\\game_data"
    ];

    let savesFound = [];
    for (const d of defaultDirs) {
        if (fs.existsSync(d)) {
            const files = fs.readdirSync(d).filter(f => f.endsWith('.es3') && !f.toLowerCase().includes('index'));
            for (const f of files) {
                savesFound.push(path.join(d, f));
            }
        }
    }

    let savePath = "";
    const isAuto = process.argv.includes('--auto');

    if (savesFound.length > 0) {
        if (isAuto) {
            savePath = savesFound[0];
            console.log(`\n[Auto] Using automatically found save: ${savePath}`);
        } else {
            console.log("\nFound save games:");
            savesFound.forEach((s, idx) => {
                console.log(` [${idx + 1}] ${s}`);
            });
            console.log(" [M] Manual path entry");

            const choice = (await question("\nSelect a save file to edit: ")).trim();
            if (choice.toLowerCase() === 'm') {
                savePath = (await question("Enter full path to save file (.es3): ")).trim().replace(/^"|"$/g, '');
            } else {
                const idx = parseInt(choice) - 1;
                if (idx >= 0 && idx < savesFound.length) {
                    savePath = savesFound[idx];
                } else {
                    console.log("Invalid choice.");
                    rl.close();
                    return;
                }
            }
        }
    } else {
        if (isAuto) {
            console.log("[-] Auto mode failed: No save games found automatically.");
            rl.close();
            return;
        }
        savePath = (await question("\nNo saves found automatically. Enter full path to save file (.es3): ")).trim().replace(/^"|"$/g, '');
    }

    if (!savePath || !fs.existsSync(savePath)) {
        console.log("Invalid file path or file does not exist.");
        rl.close();
        return;
    }

    // Backup
    const backupPath = savePath + ".bak";
    if (!fs.existsSync(backupPath)) {
        fs.copyFileSync(savePath, backupPath);
        console.log(`\n[+] Created backup of original save file at: ${backupPath}`);
    }

    // Read and decrypt
    console.log("\nDecrypting save file...");
    const rawData = fs.readFileSync(savePath);
    const decResult = decrypt_es3(rawData, PASSWORD);

    if (!decResult.success) {
        console.log(`[-] Decryption failed: ${decResult.error}`);
        rl.close();
        return;
    }

    let saveJSON;
    const cleanedText = preprocess_es3_json(decResult.text);
    try {
        saveJSON = JSON.parse(cleanedText);
    } catch (e) {
        console.log("[-] Decrypted content is not valid JSON. Save file might be corrupted or using a different password.");
        console.log(decResult.text.substring(0, 100));
        rl.close();
        return;
    }

    console.log("[+] Decrypted successfully!");

    let targetObj = saveJSON;
    if (saveJSON.key && typeof saveJSON.key.value === 'object' && saveJSON.key.value !== null) {
        targetObj = saveJSON.key.value;
    }

    // Show some keys
    console.log("\n--- Save File Keys & Values ---");
    const keys = Object.keys(targetObj);
    keys.slice(0, 15).forEach(k => {
        console.log(` - ${k}: ${JSON.stringify(targetObj[k])}`);
    });
    if (keys.length > 15) {
        console.log(` ... and ${keys.length - 15} more keys.`);
    }

    let menuChoice = "";
    let modified = false;

    if (isAuto) {
        console.log("\n[Auto] Modifying save data...");
        // Modify money
        const moneyKeys = keys.filter(k => /money|gold|cash|coin/i.test(k));
        moneyKeys.forEach(k => {
            console.log(`Changing ${k} from ${JSON.stringify(targetObj[k])} to 9999999`);
            targetObj[k] = 9999999;
        });
        
        // Modify fails
        const failKeys = keys.filter(k => /fail/i.test(k));
        failKeys.forEach(k => {
            console.log(`Changing ${k} from ${JSON.stringify(targetObj[k])} to 0`);
            targetObj[k] = 0;
        });
        modified = true;
    } else {
        console.log("\n=== Editing Menu ===");
        console.log(" [1] Set Money to 9,999,999");
        console.log(" [2] Set task_fail_count to 0 (Reset mission failures)");
        console.log(" [3] Custom JSON Modification (advanced)");
        console.log(" [4] Just Export Decrypted JSON to text file");

        menuChoice = (await question("\nChoose an option: ")).trim();

        if (menuChoice === "1") {
            const moneyKeys = keys.filter(k => /money|gold|cash|coin/i.test(k));
            if (moneyKeys.length > 0) {
                moneyKeys.forEach(k => {
                    console.log(`Changing ${k} from ${JSON.stringify(targetObj[k])} to 9999999`);
                    targetObj[k] = 9999999;
                });
                modified = true;
            } else {
                const customKey = (await question("Money key not found automatically. Enter key name (e.g. 'coin' or 'money'): ")).trim();
                if (customKey in targetObj) {
                    targetObj[customKey] = 9999999;
                    modified = true;
                } else {
                    console.log(`Key '${customKey}' not found in save.`);
                }
            }
        } else if (menuChoice === "2") {
            const failKeys = keys.filter(k => /fail/i.test(k));
            if (failKeys.length > 0) {
                failKeys.forEach(k => {
                    console.log(`Changing ${k} from ${JSON.stringify(targetObj[k])} to 0`);
                    targetObj[k] = 0;
                });
                modified = true;
            } else {
                const customKey = (await question("Failure key not found. Enter key name (e.g. 'task_fail_count'): ")).trim();
                if (customKey in targetObj) {
                    targetObj[customKey] = 0;
                    modified = true;
                } else {
                    console.log(`Key '${customKey}' not found.`);
                }
            }
        } else if (menuChoice === "3") {
            const customKey = (await question("Enter the key name to modify: ")).trim();
            if (customKey in targetObj) {
                console.log(`Current value: ${JSON.stringify(targetObj[customKey])} (type: ${typeof targetObj[customKey]})`);
                let newVal = (await question("Enter new value: ")).trim();
                
                if (newVal.toLowerCase() === 'true') newVal = true;
                else if (newVal.toLowerCase() === 'false') newVal = false;
                else if (!isNaN(newVal)) {
                    newVal = newVal.includes('.') ? parseFloat(newVal) : parseInt(newVal);
                }
                
                targetObj[customKey] = newVal;
                modified = true;
            } else {
                console.log(`Key '${customKey}' not found.`);
            }
        } else if (menuChoice === "4") {
            const txtPath = savePath + "_decrypted.json";
            fs.writeFileSync(txtPath, JSON.stringify(saveJSON, null, 4), 'utf8');
            console.log(`[+] Decrypted JSON exported to: ${txtPath}`);
            console.log("You can edit this file manually and then run this script again to re-encrypt it.");
        }
    }

    if (modified) {
        console.log("\nRe-encrypting and saving file...");
        const jsonStr = postprocess_es3_json(JSON.stringify(saveJSON));
        const newEncrypted = encrypt_es3(jsonStr, PASSWORD, decResult.wasGunzipped);
        fs.writeFileSync(savePath, newEncrypted);
        console.log("[+] Save file updated successfully! Enjoy!");
    }

    rl.close();
}

main();
